index/segment/mem/segment.go

```diff
@@ -22,224 +22,250 @@ package mem
+		offset:    int(offset),
 		newUUIDFn: opts.NewUUIDFn(),
 		termsDict: newTermsDict(opts),
+		readerID:  postings.NewAtomicID(offset),
 	}
 
 	s.docs.data = make([]doc.Document, opts.InitialCapacity())
 
-	s.ids.writer = postings.NewAtomicID(offset)
-	s.ids.reader = postings.NewAtomicID(offset)
+	s.writer.idSet = idsgen.New(256)
+	s.writer.nextID = offset
 	return s, nil
 }
 
-func (s *segment) Insert(d doc.Document) error {
+func (s *segment) Insert(d doc.Document) ([]byte, error) {
 	s.state.RLock()
 	defer s.state.RUnlock()
 	if s.state.closed {
-		return sgmt.ErrClosed
-	}
-
-	if s.state.sealed {
-		return errSegmentSealed
+		return nil, sgmt.ErrClosed
 	}
 
-	docID, err := d.Validate()
-	if err != nil {
-		return err
-	}
+	{
+		s.writer.Lock()
 
-	if docID != nil {
-		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
-		if err != nil {
-			return fmt.Errorf("could not lookup document ID '%s'", docID)
-		}
-		if ok {
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
+		ds := []doc.Document{d}
+		if err := s.prepareDocsWithLocks(ds); err != nil {
+			return nil, err
 		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
 
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+		// Update the document in case we generated a UUID for it.
+		d = ds[0]
 
-	newID := s.ids.writer.Inc()
-	s.insertDoc(newID, d)
-	err = s.insertTerms(newID, d)
-	s.ids.reader.Inc()
+		s.insertDocWithLocks(d)
+		s.readerID.Inc()
 
-	s.ids.Unlock()
+		s.writer.Unlock()
+	}
 
-	return err
+	return d.ID, nil
 }
 
-func (s *segment) Reader() (index.Reader, error) {
+func (s *segment) InsertBatch(ds []doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
-		return nil, sgmt.ErrClosed
+		return sgmt.ErrClosed
 	}
 
-	maxID := s.ids.reader.Load()
-	r := newReader(s, maxID)
-	return r, nil
-}
+	{
+		s.writer.Lock()
 
-func (s *segment) Seal() error {
-	s.state.Lock()
-	if s.state.sealed {
-		s.state.Unlock()
-		return errSegmentSealed
+		if err := s.prepareDocsWithLocks(ds); err != nil {
+			return err
+		}
+
+		for _, d := range ds {
+			s.insertDocWithLocks(d)
+		}
+		s.readerID.Add(uint32(len(ds)))
+
+		s.writer.Unlock()
 	}
 
-	s.state.sealed = true
-	s.state.Unlock()
 	return nil
 }
 
-func (s *segment) Close() error {
-	s.state.Lock()
-	if s.state.closed {
-		s.state.Unlock()
-		return sgmt.ErrClosed
+// prepareDocsWithLocks ensures the given documents can be inserted into the index. It
+// must be called with the state and writer locks.
+func (s *segment) prepareDocsWithLocks(ds []doc.Document) error {
+	s.writer.idSet.Reset()
+	for i := 0; i < len(ds); {
+		d := ds[i]
+		err := d.Validate()
+		if err != nil {
+			return err
```

Comment:
instead of early terminating, could you continue attempting to index the remainder of the docs